File: 23. Merge k Sorted Lists/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        # create a heap of size k - number of lists
        heap = []
        for l in lists:
            if l:
                heapq.heappush(heap, (l.val, l))
            
        # extract and pushback
        head = ListNode(None)
        node = head
        while heap:
            _, l = heapq.heappop(heap)
            node.next = l
            node = node.next
            l = l.next
            node.next = None
            if l:
                heapq.heappush(heap, (l.val, l))
                
        return head.next

File: 23. Merge k Sorted Lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        ListNode head = new ListNode(0);
        ListNode node = head;
        int low = 0;
        int min;
        while (true) {
            low = -1;
            min = Integer.MAX_VALUE;
            for(int i = 0; i < lists.length; i++) {
                if (lists[i] != null && lists[i].val < min) {
                    low = i;
                    min = lists[i].val;
                }
            }
            if (low == -1) {
                return head.next;
            }
            node.next = lists[low];
            lists[low] = lists[low].next;
            node = node.next;
        }
    }
}
